### PR TITLE
feat(pivot): RS, SF, BQ

### PR DESCRIPTION
**Pull request by raphaelvignes**
Implement a **static** pivot step for:
* Snowflake
* Redshift
* BigQuery

### CODE REVIEW

**Comment by raphaelvignes** [comment]
Capture d’écran du 2022-08-19 15-15-42

**Comment by raphaelvignes** [comment]
TODO: Pimp the multi input field. 
* Add a title
* Sanitize the input (handle strings with whitespace)
image

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 1]
```diff
@@ -1119,6 +1121,21 @@ def percentage(
     ) -> StepContext:
         raise NotImplementedError(f"[{self.DIALECT}] percentage is not implemented")
 
+    @classmethod
+    def _build_pivot_col(
+        cls,
+        *,
+        step: 'PivotStep',
+        quote_char: str | None,
+        secondary_quote_char: str,
```
Comment (Sanix-Darker) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:1130]:
```suggestion
        secondary_quote_char: str | None
```

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 2]
```diff
@@ -1119,6 +1121,21 @@ def percentage(
...
+        prev_step_name: str | None,
+    ) -> str:
+        pivot_col = format_quotes(step.column_to_pivot, quote_char)
+        aggregated_col = f'{step.agg_function}({format_quotes(step.value_column, quote_char)})'
+        assert step.values
```
Comment (Sanix-Darker) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:1135]:
```suggestion
        assert step.values, , "step.values is missing from this pivot step"
```

File: server/src/weaverbird/backends/pypika_translator/translators/googlebigquery.py [hunk 3]
```diff
@@ -63,6 +65,27 @@ def _add_date(
...
+    ) -> StepContext:
+        if not (self.SUPPORT_PIVOT):
+            raise NotImplementedError(f"[{self.DIALECT}] pivot is not implemented")
+        pivot = self._build_pivot_col(
+            step=step,
+            quote_char=builder.QUOTE_CHAR,
+            secondary_quote_char=builder.SECONDARY_QUOTE_CHAR,
+            prev_step_name=None,
+        )
+        assert step.values
```
Comment (Sanix-Darker) [inline server/src/weaverbird/backends/pypika_translator/translators/googlebigquery.py:84]:
```suggestion
        assert step.values, "step.values is missing from this pivot step"
```

File: server/src/weaverbird/backends/pypika_translator/translators/redshift.py [hunk 4]
```diff
@@ -85,5 +87,48 @@ def concatenate(
...
+        aggregated_col = f'{step.agg_function}({format_quotes(step.value_column, quote_char)})'
+        assert step.values
+        value_cols = ', '.join(format_quotes(val, secondary_quote_char) for val in step.values)
+        selected_cols = (
+            ', '.join(step.index)
+            + ', '
+            + pivot_col
+            + ', '
+            + format_quotes(step.value_column, quote_char)
+        )
```
Comment (Sanix-Darker) [inline server/src/weaverbird/backends/pypika_translator/translators/redshift.py:109]:
```suggestion
        selected_cols = (
            f'{", ".join(step.index)}, {pivot_col}, {format_quotes(step.value_column, quote_char)}'
        )
```

File: server/src/weaverbird/backends/pypika_translator/translators/snowflake.py [hunk 5]
```diff
@@ -91,5 +99,35 @@ def _get_date_extract_func(cls, *, date_unit: DATE_INFO, target_column: Field) -
...
+    ) -> StepContext:
+        if not (self.SUPPORT_PIVOT):
+            raise NotImplementedError(f"[{self.DIALECT}] pivot is not implemented")
+        pivot = self._build_pivot_col(
+            step=step,
+            quote_char=builder.QUOTE_CHAR,
+            secondary_quote_char=builder.SECONDARY_QUOTE_CHAR,
+            prev_step_name=None,
+        )
+        assert step.values
```
Comment (Sanix-Darker) [inline server/src/weaverbird/backends/pypika_translator/translators/snowflake.py:118]:
```suggestion
        assert step.values, "step.values is missing from this pivot step"
```

**Comment by davinov** [comment]
Closing this PR, stale for a long time

**Pull request closed without merging**